zmap/zgrab2
# Add NTP zgrab2 module

NTP module for zgrab2. Supports both normal operation and sending MONLIST requests.

## How to Test

Normal operation can be tested against any NTP server, e.g. `echo "time.nist.gov" | cmd/zgrab2/zgrab2 ntp --timeout 3 | jp data`. The timestamp should appear in `data.ntp.result.time`.

To test the `--monlist` mode, you obviously will need an NTP service with MONLIST enabled -- the container in `integration-test/ntp/container-4.2.6` can serve this purpose (it builds [NTP 4.2.6](https://www.eecis.udel.edu/~ntp/ntp_spool/ntp4/ntp-4.2/ntp-4.2.6p5.tar.gz) with the default options, which allows MONLIST). So you could run

 * `docker build -t zgrab_ntp:4.2.6 integration_tests/ntp/container-4.2.6`
 * `docker run -itd -p 33123:123/udp --rm --name ntptest zgrab_ntp:4.2.6`
 * `echo "127.0.0.1" | cmd/zgrab2/zgrab2 ntp --port 33123 --timeout 3 --monlist`

There should be a `data.ntp.result.monlist_response` with some binary data and a `data.ntp.result.monlist_header` with more details on the response, in addition to the `data.ntp.result.time` and `data.ntp.result.time_response` fields that were present before (they should disappear if you pass `--skip-get-time`).

## Notes & Caveats

 * For some reason, on my machine at least, forwarding UDP to ::1 on docker containers does not seem to work (forwarding to IPv4 addresses does, and forwarding TCP to IPv6 addresses works as well).

## Issue Tracking

Part of the DDoS amplifiers list
